src/NuGet.Core/NuGet.Packaging/Core/FrameworkNameValidatorUtility.cs

```diff
@@ -36,25 +36,30 @@ internal static bool IsValidFrameworkName(string path)
 
         internal static bool IsDottedFrameworkVersion(string path)
         {
-            foreach (string knownFolder in PackagingConstants.Folders.Known)
+            foreach (string knownFolder in PackagingConstants.Folders.SupportFrameworks)
             {
                 string folderPrefix = knownFolder + System.IO.Path.DirectorySeparatorChar;
-                if (knownFolder == "contentFiles")
-                {
-                    folderPrefix += "any" + System.IO.Path.DirectorySeparatorChar;
-                }
                 if (path.Length > folderPrefix.Length &&
                     path.StartsWith(folderPrefix, StringComparison.OrdinalIgnoreCase))
                 {
-                    string frameworkPart = path.Substring(folderPrefix.Length);
+                    string frameworkPart = null;
+                    if (knownFolder == PackagingConstants.Folders.ContentFiles)
+                    {
+                        // ContentFiles has an extra level of nesting for the
+                        // language, so we have to skip over _that_ in order
+                        // to get to the framework itself.
+                        var languageEndIndex = path.IndexOf(System.IO.Path.DirectorySeparatorChar, folderPrefix.Length + 1);
+                        frameworkPart = path.Substring(folderPrefix.Length + (folderPrefix.Length - languageEndIndex));
+                    }
+                    else
+                    {
+                        frameworkPart = path.Substring(folderPrefix.Length);
```

Comment:
I just remembered that a class named `ManagedCodeConventions` exists, and it lists the path conventions for each. For example, look at the runtimes folder:  https://github.com/NuGet/NuGet.Client/blob/9f521755495c05bc03efd5365762e32fde98761f/src/NuGet.Core/NuGet.Packaging/ContentModel/ManagedCodeConventions.cs#L415

also the multi-targetd build files has a deprecated path that is still used:
https://github.com/NuGet/NuGet.Client/blob/9f521755495c05bc03efd5365762e32fde98761f/src/NuGet.Core/NuGet.Packaging/ContentModel/ManagedCodeConventions.cs#L505-L508

I assume that the `PatternSet` class has methods to parse the path for you, so you don't need to do all this substring stuff.

I'm really sorry I didn't remember this earlier :(  particularly as the current implementation doesn't correctly get the TFM for all package folders.

Reply:
I even remember giving this feedback to Joshua a while ago :( 

Here's a rule that makes use of that. 

https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Packaging/Rules/NoRefOrLibFolderInPackageRule.cs